go/vt/vttablet/tabletserver/throttle/throttler.go

```diff
@@ -394,105 +506,115 @@ func (throttler *Throttler) isDormant() bool {
+	go func() {
+		defer log.Infof("Throttler: Operate terminated, tickers stopped")
+		for _, t := range tickers {
+			defer t.Stop()
+			// since we just started the tickers now, speed up the ticks by forcng an immediate tick
+			go t.TickNow()
+		}
 
-					// sparse
-					shouldBeLeader := int64(0)
-					if atomic.LoadInt64(&throttler.isOpen) > 0 {
-						if throttler.tabletTypeFunc() == topodatapb.TabletType_PRIMARY {
-							shouldBeLeader = 1
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-leaderCheckTicker.C:
+				{
+					func() {
+						throttler.initMutex.Lock()
+						defer throttler.initMutex.Unlock()
+
+						// sparse
+						shouldBeLeader := int64(0)
+						if atomic.LoadInt64(&throttler.isOpen) > 0 {
+							if throttler.tabletTypeFunc() == topodatapb.TabletType_PRIMARY {
+								shouldBeLeader = 1
+							}
 						}
-					}
 
-					transitionedIntoLeader := false
-					if shouldBeLeader > throttler.isLeader {
-						log.Infof("Throttler: transition into leadership")
-						transitionedIntoLeader = true
-					}
-					if shouldBeLeader < throttler.isLeader {
-						log.Infof("Throttler: transition out of leadership")
-					}
+						transitionedIntoLeader := false
+						if shouldBeLeader > throttler.isLeader {
+							log.Infof("Throttler: transition into leadership")
+							transitionedIntoLeader = true
+						}
+						if shouldBeLeader < throttler.isLeader {
+							log.Infof("Throttler: transition out of leadership")
+						}
 
-					atomic.StoreInt64(&throttler.isLeader, shouldBeLeader)
+						atomic.StoreInt64(&throttler.isLeader, shouldBeLeader)
 
-					if transitionedIntoLeader {
-						// transitioned into leadership, let's speed up the next 'refresh' and 'collect' ticks
-						go mysqlRefreshTicker.TickNow()
-						go throttler.heartbeatWriter.RequestHeartbeats()
+						if transitionedIntoLeader {
+							// transitioned into leadership, let's speed up the next 'refresh' and 'collect' ticks
+							go mysqlRefreshTicker.TickNow()
+							go throttler.heartbeatWriter.RequestHeartbeats()
+						}
+					}()
+				}
+			case <-mysqlCollectTicker.C:
+				{
+					if atomic.LoadInt64(&throttler.isOpen) > 0 {
+						// frequent
+						if !throttler.isDormant() {
+							throttler.collectMySQLMetrics(ctx)
+						}
 					}
-				}()
-			}
-		case <-mysqlCollectTicker.C:
-			{
-				if atomic.LoadInt64(&throttler.isOpen) > 0 {
-					// frequent
-					if !throttler.isDormant() {
-						throttler.collectMySQLMetrics(ctx)
+				}
+			case <-mysqlDormantCollectTicker.C:
+				{
+					if atomic.LoadInt64(&throttler.isOpen) > 0 {
+						// infrequent
+						if throttler.isDormant() {
+							throttler.collectMySQLMetrics(ctx)
+						}
 					}
 				}
-			}
-		case <-mysqlDormantCollectTicker.C:
-			{
-				if atomic.LoadInt64(&throttler.isOpen) > 0 {
-					// infrequent
-					if throttler.isDormant() {
-						throttler.collectMySQLMetrics(ctx)
+			case metric := <-throttler.mysqlThrottleMetricChan:
```

Comment:
Just noting that these could be skipped if other things fire concurrently. Assuming that's OK, even if it's more than one consecutively?

Reply:
not sure how these would be skipped? There will be some `golang` scheduling to determine who gets to be executed first, but these events won't be skipped. At most, they will wait a bit, which is fine. There is nothing that hard-spams this `select` clause.

Reply:
Sorry, I meant skipped in the loop. Sounds like that's fine.

Reply:
I meant skipped in the for loop due to there being so many clauses (from the [spec](https://go.dev/ref/spec#Select_statements)):
```
If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection. Otherwise, if there is a default case, that case is chosen. If there is no default case, the "select" statement blocks until at least one of the communications can proceed.
```

Sounds like it's fine.

Reply:
Right - it's fine, there's no problem is some other clause grabs attention a few times before this one does.